reference_controllers/rabbitmq_controller_unverified/src/main.rs

```diff
@@ -0,0 +1,340 @@
+#![allow(clippy::unnecessary_lazy_evaluations)]
+#![allow(unused_imports)]
+
+pub mod default_user_secret;
+pub mod erlang_cookie;
+pub mod headless_service;
+pub mod rabbitmq_plugins;
+pub mod rabbitmqcluster_status;
+pub mod rabbitmqcluster_types;
+pub mod role;
+pub mod role_binding;
+pub mod server_configmap;
+pub mod service;
+pub mod service_account;
+pub mod statefulset;
+
+use anyhow::Result;
+use core::time;
+use futures::StreamExt;
+use k8s_openapi::api::core::v1 as corev1;
+use k8s_openapi::api::rbac::v1 as rbacv1;
+use k8s_openapi::api::{apps::v1 as appsv1, core::v1::Service};
+use kube::{
+    api::{Api, ListParams, PostParams},
+    client::ConfigExt,
+    runtime::controller::{Action, Controller},
+    Client, Config, CustomResourceExt,
+};
+use kube_client::{self, client};
+use kube_core::{self, ResourceExt};
+use std::str::FromStr;
+use std::{env, sync::Arc};
+use thiserror::Error;
+use tokio::time::Duration;
+use tracing::*;
+
+use crate::rabbitmqcluster_types::*;
+
+#[derive(Debug, Error)]
+enum Error {
+    // #[error("Failed to get CR: {0}")]
+    // CRGetFailed(#[source] kube::Error),
+    #[error("Failed to create ConfigMap: {0}")]
+    ConfigMapCreationFailed(#[source] kube::Error),
+
+    #[error("Failed to create Service: {0}")]
+    ServiceCreationFailed(#[source] kube::Error),
+    #[error("Failed to create ServiceAccount: {0}")]
+    ServiceAccountCreationFailed(#[source] kube::Error),
+
+    #[error("Failed to create Role: {0}")]
+    RoleCreationFailed(#[source] kube::Error),
+    #[error("Failed to create RoleBinding: {0}")]
+    RoleBindingCreationFailed(#[source] kube::Error),
+
+    #[error("Failed to create StatefulSet: {0}")]
+    StatefulSetCreationFailed(#[source] kube::Error),
+
+    // #[error("MissingObjectKey: {0}")]
+    // MissingObjectKey(&'static str),
+    #[error("ReplaceImageFail: {0}")]
+    ReplaceImageFail(kube_client::Error),
+}
+
+struct RabbitmqClusterReconciler {
```

Comment:
I would recommend you not call it `RabbitmqClusterReconciler` because `reconciler` has a slightly different meaning in our codebase. You can just name it data or context.